aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,266 @@
+use std::sync::Arc;
+
+#[allow(unused_imports)]
+use crate::aggregator::Config as AggregatorConfig; // used in doccomment.
+use anyhow::{anyhow, Context, Error};
+use janus_aggregator_core::datastore::{
+    models::{GlobalHpkeKeypair, HpkeKeyState},
+    Datastore, Error as DatastoreError, Transaction,
+};
+use janus_core::{
+    hpke::{generate_hpke_config_and_private_key, HpkeCiphersuite, HpkeKeypair},
+    time::{Clock, TimeExt},
+};
+use janus_messages::{Duration, HpkeConfigId};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug)]
+pub struct KeyRotator<C: Clock> {
+    clock: C,
+    datastore: Arc<Datastore<C>>,
+    hpke: HpkeKeyRotatorConfig,
+}
+
+/// Handles key rotation for global HPKE keys. It is assumed that the [`KeyRotator`] is the sole
+/// entity responsible for the `global_hpke_keys` table. Do not make manual changes to the
+/// `global_hpke_keys` table when the key rotator is used.
```

Comment:
I think we'll need to be able to support manually-initiated key rotations, (informed by our experience with ENPA) even if it's a multi-step process. If we make the key rotator robust to multiple pending keys and multiple active keys, that will help out a lot. We will still have to tell the user not to delete active keys, and not to promote pending keys to active early unless they've rebooted aggregator processes, and write up a playbook with janus_cli commands for a manual rotation.

Previously, with `key-rotator`, we had to emulate manual rotations by changing times in `key-rotator`'s configuration. That led to some surprises when we did so, and when we restored the original configuration again. I previously editorialized about this here: https://github.com/divviup/janus-ops/issues/142#issuecomment-1629571706

One nice thing about HPKE keys for DAP, as opposed to other secrets we need to rotate, is that we can get away with having multiple "active" keys with the same ciphersuites at once, as that just means we'll have a longer HpkeConfigList. Thus, it's okay for a manual rotation playbook to update key states in separate steps, not transactionally, and we don't necessarily need a new janus_cli command.

Reply:
Thanks for the context, I'll add support for manual key rotations.